trestle/core/profile_resolver.py

```diff
@@ -239,12 +319,121 @@ def _merge_lists(
+            controls = []
+            params = []
+            if group.controls is not None:
+                controls.extend(group.controls)
+            if group.params is not None:
+                params.extend(group.params)
+            if group.groups is not None:
+                for sub_group in group.groups:
+                    new_controls, new_params = self._group_contents(sub_group)
+                    controls.extend(new_controls)
+                    params.extend(new_params)
+            return controls, params
+
+        def _flatten_catalog(self, catalog: cat.Catalog, as_is: bool) -> cat.Catalog:
+            """Flatten the groups of the catalog if as_is is False."""
+            if as_is or catalog.groups is None:
+                return catalog
+            catalog.controls = as_list(catalog.controls)
+            catalog.params = as_list(catalog.params)
+            for group in catalog.groups:
+                new_controls, new_params = self._group_contents(group)
+                catalog.controls.extend(new_controls)
+                catalog.params.extend(new_params)
+            catalog.controls = none_if_empty(catalog.controls)
+            catalog.params = none_if_empty(catalog.params)
+            catalog.groups = None
+            return catalog
 
-        def _resolve_merge_method(self) -> prof.Method:
-            """Resolve the merge method based on trestle defaults and available content."""
+        def _merge_two_catalogs(
+            self, dest: cat.Catalog, src: cat.Catalog, merge_method: prof.Method, as_is: bool
+        ) -> cat.Catalog:
+            # merge_method is use_first, merge, or keep
+            # if as_is is false, the result is flattened
+
+            dest = self._flatten_catalog(dest, as_is)
+            src = self._flatten_catalog(src, as_is)
+
+            if src.controls is not None:
+                dest.controls = as_list(dest.controls)
+                dest_ids = [control.id for control in dest.controls]
+                for control in src.controls:
+                    in_list = control.id in dest_ids
+                    if merge_method == prof.Method.keep or not in_list:
+                        dest.controls.append(control)
+                    elif merge_method == prof.Method.merge:
+                        index = dest_ids.index(control.id)
+                        self._merge_controls(dest.controls[index], control)
+
+            if src.params is not None:
+                dest.params = as_list(dest.params)
+                dest_ids = [param.id for param in dest.params]
+                for param in src.params:
+                    in_list = param.id in dest_ids
+                    if merge_method == prof.Method.keep or not in_list:
+                        dest.params.append(param)
+                    elif merge_method == prof.Method.merge:
+                        index = dest_ids.index(param.id)
+                        self._merge_params(dest.params[index], param)
```

Comment:
same lines as 387-395, could those be unified? i.e. the function on 406 could be passed as argument

Reply:
Again I'm manipulating different structures so maybe you can tell me what you had in mind.  I see that the functions can be passed in, but it's the internals I'm concerned about.